test/test_download.py

```diff
@@ -254,12 +276,49 @@ def try_rm_tcs_files(tcs=None):
         tname = 'test_%s_%d' % (test_case['name'], i)
         i += 1
     test_method = generator(test_case, tname)
-    test_method.__name__ = str(tname)
-    ie_list = test_case.get('add_ie')
-    test_method.add_ie = ie_list and ','.join(ie_list)
-    setattr(TestDownload, test_method.__name__, test_method)
-    del test_method
-
+    ie_list = ','.join(test_case.get('add_ie', []))
+    TestDownload.addTest(test_method, tname, ie_list)
+
+# Py2 compat (should be in compat.py?)
+try:
+    from itertools import (ifilter as filter, imap as map)
+except ImportError:
+    pass
+
+
+def tests_for_ie(ie_key):
+    return filter(
+        lambda a: callable(getattr(TestDownload, a, None)),
+        filter(lambda a: re.match(r'test_%s(?:_\d+)?$' % ie_key, a),
+               dir(TestDownload)))
+
+
+def gen_test_suite(ie_key):
+    def test_all(self):
+        print(self)
+        suite = unittest.TestSuite(
+            map(TestDownload, tests_for_ie(ie_key)))
+        result = self.defaultTestResult()
+        suite.run(result)
+        print('Errors: %d\t Failures: %d\tSkipped: %d' %
+              tuple(map(len, (result.errors, result.failures, result.skipped))))
+        print('Expected failures: %d\tUnexpected successes: %d' %
+              tuple(map(len, (result.expectedFailures, result.unexpectedSuccesses))))
+        return result
+
+    return test_all
+
+
+for ie_key in set(
+    map(lambda x: x[0],
+        filter(
+            lambda x: callable(x[1]),
+            map(lambda a: (a[5:], getattr(TestDownload, a, None)),
+                filter(lambda t:
+                       re.match(r'test_.+(?<!(?:_all|.._\d|._\d\d|_\d\d\d))$', t),
+                       dir(TestDownload)))))):
```

Comment:
Why do we need to create the intermediate tuple?
```suggestion
for ie_key in set(
    map(lambda a: a[5:],
        filter(
            lambda x: callable(getattr(TestDownload, x, None)),
            filter(
                lambda t: re.match(r"test_.+(?<!(?:_all|.._\d|._\d\d|_\d\d\d))$", t),
                dir(TestDownload))))):
```

Reply:
Looks like the result of Lego development. It can be simplified.